hazelcast/src/main/java/com/hazelcast/map/impl/client/MapAddNearCacheEntryListenerRequest.java

```diff
@@ -61,12 +67,46 @@ protected Object newMapListener(ClientEndpoint endpoint) {
 
         @Override
         public void onInvalidate(Invalidation event) {
-            if (!endpoint.isAlive()
-                    || endpoint.getUuid().equals(event.getSourceUuid())) {
+            if (!endpoint.isAlive()) {
+                return;
+            }
+
+            event = getFilteredEventOrNull(event);
+            if (event == null) {
                 return;
             }
 
+            sendEvent(event);
+        }
+
+        protected Invalidation getFilteredEventOrNull(Invalidation event) {
+
+            if (event instanceof BatchNearCacheInvalidation) {
+                List<SingleNearCacheInvalidation> invalidations
+                        = excludeThisSourcesInvalidations((BatchNearCacheInvalidation) event);
+                if (!isEmpty(invalidations)) {
+                    return event;
+                }
+            } else if (!endpoint.getUuid().equals(event.getSourceUuid())) {
+                return event;
+            }
+
+            return null;
+        }
+
+        protected List<SingleNearCacheInvalidation> excludeThisSourcesInvalidations(BatchNearCacheInvalidation event) {
+            List<SingleNearCacheInvalidation> invalidations = event.getInvalidations();
```

Comment:
As far as I understood, this `BatchNearCacheInvalidation` can be sent to multiple `InvalidationListener` for different endpoints. So this means that it is shared. But in here we are modifying its included `SingleNearCacheInvalidation`s by removing them through iterator. So this effects others and removed `SingleNearCacheInvalidation` would be used by another `InvalidationListener` for different endpoint. 

Am I missing something here?